src/refda/exec.c

```diff
@@ -375,91 +376,260 @@ void *refda_exec_worker(void *arg)
+                    (next->tbr.callback)(next->tbr.agent, next->tbr.tbr);
+                    break;
                 }
-                if (!atomic_load(&(next->item->waiting)))
+                default:
                 {
-                    refda_exec_ctrl_finish(next->item);
+                    CACE_LOG_ERR("Unknown type of deferred callback %d", next->purpose);
                 }
-
-                refda_timeline_remove(agent->exec_timeline, tl_it);
             }
+
+            refda_timeline_remove(agent->exec_timeline, tl_it);
         }
-        else
+    }
+    else
+    {
+        CACE_LOG_DEBUG("waiting for exec event");
+        sem_wait(&(agent->execs_sem));
+    }
+
+    // execs queue may still be empty if deferred callbacks were run
+    if (refda_msgdata_queue_pop(&item, agent->execs))
+    {
+        // sentinel for end-of-input
+        const bool at_end = cace_ari_is_undefined(&(item.value));
+        if (!at_end)
+        {
+            refda_exec_exp_execset(agent, &item);
+        }
+        refda_msgdata_deinit(&item);
+        if (at_end && refda_timeline_empty_p(agent->exec_timeline))
         {
-            CACE_LOG_DEBUG("waiting for exec event");
-            sem_wait(&(agent->execs_sem));
+            CACE_LOG_INFO("Got undefined exec, stopping");
+
+            // flush the input queue but keep the daemon running
+            refda_msgdata_t undef;
+            refda_msgdata_init(&undef);
+            refda_msgdata_queue_push_move(agent->rptgs, &undef);
+            sem_post(&(agent->rptgs_sem));
+
+            return;
         }
+    }
+
+    // execute any waiting sequences
+    refda_exec_waiting(agent);
+}
+
+static int refda_exec_schedule_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr, bool starting);
+
+/** Execute a time based rule's action that has already been verified
+ * Based on code from refda_exec_exp_execset
+ */
+static int refda_exec_tbr_action(refda_agent_t *agent, refda_exec_seq_t *seq, const refda_amm_tbr_desc_t *tbr)
+{
+    refda_runctx_ptr_t ctxptr;
+    refda_runctx_ptr_init_new(ctxptr);
+
+    refda_runctx_t *runctx = refda_runctx_ptr_ref(ctxptr);
+
+    if (refda_runctx_from(runctx, agent, NULL))
+    {
+        return 2;
+    }
+
+    refda_runctx_ptr_set(seq->runctx, ctxptr);
+    int res = refda_exec_exp_mac(runctx, seq, &(tbr->action));
+
+    refda_runctx_ptr_clear(ctxptr); // Clean up extra reference created by ptr_ref
+    return res;
+}
+
+/** Begin a single execution of a time based rule
+ */
+static void refda_exec_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(agent);
+    CHKERR1(tbr);
+
+    if (!tbr->enabled)
+    {
+        CACE_LOG_INFO("TBR is not enabled");
+        return;
+    }
+
+    if (refda_amm_tbr_desc_reached_max_exec_count(tbr))
+    {
+        CACE_LOG_INFO("TBR reached maximum execution count");
+        return;
+    }
+
+    if (tbr->action.is_ref || tbr->action.as_lit.ari_type != CACE_ARI_TYPE_AC)
```

Comment:
This is late to catch this. Can this be caught earlier in the process, with an assumption added to the `action` member (in its doxygen comment) that it must be a valid exec target?

Reply:
Yes, updated the code to catch this prior to the rule being enabled, and updated the doxygen comment with your recommendation.